drivers: can: validate the use of CAN IDs and CAN ID masks before passing them to the drivers

henrikbrixandersen opened the pull request:
- Validate the CAN ID used in the can_frame struct before passing it to the drivers.
- Validate the CAN ID and CAN ID mask used in the can_filter struct before passing it to the drivers.
- Remove the use of bitfields in the can_frame struct.
- Remove the use of bitfields in the can_filter struct.
- Add tests for sending CAN frames with invalid CAN IDs.
- Add tests for adding CAN RX filters with invalid CAN IDs and CAN ID masks.

The above changes in validation revealed an issue with the CAN ISO-TP protocol implementation, which was using 29-bit CAN ID masks regardless of the type of CAN RX filter.

henrikbrixandersen committed:
drivers: can: validate arguments in can_send()

Validate the CAN ID used in the can_frame struct before passing it to the
driver.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
drivers: can: remove the use of bitfields from struct can_frame

Remove the use of bitfields in the can_frame struct.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
tests: drivers: can: api: test sending invalid CAN frames

Add tests for sending CAN frames with invalid CAN IDs.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
drivers: can: validate arguments in can_add_rx_filter()

Validate the CAN ID and CAN ID mask used in the can_filter struct before
passing it to the driver.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
drivers: can: remove the use of bitfields from struct can_filter

Remove the use of bitfields in the can_filter struct.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
tests: drivers: can: api: test adding invalid CAN RX filters

Add tests for adding CAN RX filters with invalid CAN IDs and CAN ID masks.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
canbus: isotp: use proper CAN ID masks

Use CAN ID masks matching the flags used in the CAN RX filter.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen commented:
@str4t0m, @alexanderwachter, @martinjaeger I could use some help in reviewing this.

str4t0m reviewed (dismissed):
Sorry for the long response time. Looks all fine to me.

martinjaeger commented:
Generally looks good. However, I was wondering what the advantage of the new approach is. With bitfields one would get some of the checks at compile-time instead of runtime.

henrikbrixandersen commented:
> Generally looks good. However, I was wondering what the advantage of the new approach is. With bitfields one would get some of the checks at compile-time instead of runtime.

The bitfields only allow for checking against 29-bit CAN IDs, so they're only half the solution. One would still have to check the flags and check 11-bit CAN IDs.

Furthermore, it is possible to assign a larger value to the bitfield in code. This will result in a compiler warning and a wrap-around, but again - it only does this for 29-bit CAN IDs. 

As you can tell from the PR description, these changes actually caught a few issues in the ISO-TP implementation.

henrikbrixandersen commented on drivers/can/can_common.c:
```diff
@@ -35,6 +35,21 @@ int z_impl_can_send(const struct device *dev, const struct can_frame *frame,
+		id_mask = CAN_STD_ID_MASK;
+	}
+
+	if ((frame->id & ~(id_mask)) != 0U) {
```
I am considering changing these to use the `CHECKIF()` macro instead, allowing users to turn off these runtime checks if desired. Opinions?

martinjaeger replied:
Yeah, thought about using `__ASSERT` instead of normal `if`s, but `CHECKIF()` is probably the best compromise. I am always a bit wary of adding too many runtime checks even for things which are quite obvious errors and could thus be detected during debugging (e.g. in debug-build-only checks).

We are using CAN also on low-cost STM32F0 series MCUs, where a small-footprint driver implementation is really important.

henrikbrixandersen replied:
That's fair. I've moved this PR back into draft state until I have reimplemented it using `CHECKIF()`.

henrikbrixandersen replied:
I have updated this PR to use `CHECKIF()` for the arguments validation, allowing users to disable these run-time checks.

str4t0m replied:
This was a good idea.

henrikbrixandersen committed:
drivers: can: remove unnecessary asserts

Remove unnecessary asserts from various CAN controller drivers. These
asserts are all covered by the common CAN subsystem API implementations.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

martinjaeger reviewed (approved):
Thanks for the update to use `CHECKIF`.
